crates/compiler/src/pass/mod.rs

```diff
@@ -0,0 +1,370 @@
+//!
+//! Every Pass should implement [`ConcretePass`], as this provides the full set
+//! of features required of a pass. It is also expected that each pass provide a
+//! type-specific constructor function called `new`.
+//!
+//! This compiler will take advantage of passes defined in LLVM—that we can use
+//! via [`inkwell`]—and also custom passes tailored to the `CairoVM` CPU model
+//! that may depend on LLVM-defined analyses.
+//!
+//! # Passes
+//!
+//! A pass is a self-contained unit of functionality that performs some
+//! operation over the LLVM IR. They tend to fall into the following three
+//! categories:
+//!
+//! - **Analysis:** These examine the structure of the IR to infer information
+//!   about it without changing its structure. The information produced by
+//!   analysis passes can be invalidated by transformation passes.
+//! - **Transformation:** Transformation passes use either information from
+//!   analysis passes or structural information about the IR to change the
+//!   structure of the IR. These structural changes can happen for many reasons,
+//!   but usually involve optimizing some metric (e.g. runtime or code size).
+//! - **Utility:** These passes do not fall neatly into either of the above
+//!   categories.
+//!
+//! # Note: Skeleton
+//!
+//! The implementations in this file are deliberately left incomplete, and exist
+//! only as skeletons to serve the purposes of correctly designing the compiler
+//! state. A proper implementation will take place later in the project, as
+//! tracked by [#56](https://github.com/reilabs/llvm-to-cairo/issues/56).
+
+pub mod analysis;
+pub mod data;
+
+use std::{
+    any::{Any, TypeId},
+    fmt::Debug,
+};
+
+use derivative::Derivative;
+use downcast_rs::Downcast;
+use ltc_errors::compile::{Error, Result};
+
+use crate::{
+    pass::data::{ConcretePassData, DynPassDataMap, PassData},
+    source::SourceContext,
+};
+
+/// A pass is a self-contained unit of functionality that performs some
+/// operation over the LLVM IR.
+pub type Pass = Box<dyn PassOps>;
+
+/// A handle that uniquely identifies the pass.
+pub type PassKey = TypeId;
+
+/// Pass return data that returns a dynamic [`PassData`].
+pub type DynPassReturnData = PassReturnData<PassData>;
+
+/// The data returned when executing a pass.
+#[derive(Derivative)]
+#[derivative(Debug(bound = "T: Debug"))]
+pub struct PassReturnData<T> {
+    /// The newly-modified source context.
+    pub source_context: SourceContext,
+
+    /// The data returned by the pass.
+    pub data: T,
+}
+impl<T> PassReturnData<T> {
+    /// Creates a new instance of the pass return data.
+    pub fn new(source_context: SourceContext, data: T) -> Self {
+        Self {
+            source_context,
+            data,
+        }
+    }
+}
+
+impl PassReturnData<PassData> {
+    /// Allows you to get the returned pass data as the concrete data type `T`,
+    /// returning `&T` if possible and `None` otherwise.
+    #[must_use]
```

Comment:
If you mean the decision to use it, it is good practice to annotate these as `#[must_use]` because if they call it and _don't_ use it then it's probably a mistake.